Fixes to apiserver for subresources and naming

smarterclayton opened the pull request:
@derekwaynecarr @liggitt review please

smarterclayton committed:
Verify namespace is not set on root objects

smarterclayton committed:
Subresources should be in their parent rest scope

A subresource like "Binding" does not necessarily have
to have a namespace.  The RESTScope of a subresource
should always be its parent resource.

smarterclayton committed:
Legacy scope naming should NEVER set namespace for root

... resources.

liggitt commented on pkg/apiserver/api_installer.go:
```diff
@@ -128,6 +129,25 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
+			return err
+		}
+		mapping.Scope = parentMapping.Scope
+		log.Printf("set scope %s %#v", kind, mapping.Scope)
```
remove log

liggitt commented on pkg/apiserver/api_installer.go:
```diff
@@ -128,6 +129,25 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 
+	// subresources must have parent resources, and follow the namespacing rules of their parent
+	if hasSubresource {
+		storage, ok := a.group.Storage[resource]
```
call it `parentStorage`, `parentKind`, and `parentObject` for sanity?

liggitt commented on pkg/apiserver/api_installer.go:
```diff
@@ -128,6 +129,25 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
+		if err != nil {
+			return err
+		}
+		mapping.Scope = parentMapping.Scope
```
if the scopes aren't equal, should we err rather than stomp?

smarterclayton replied:
No - the scope of the child is immaterial in a sub resource.  Sub resources detach scope from kind - the scope is "the scope of my parent" always.

liggitt commented on pkg/apiserver/api_installer.go:
```diff
@@ -689,6 +709,9 @@ var _ ScopeNamer = legacyScopeNaming{}
 
 // Namespace returns the namespace from the query or the default.
 func (n legacyScopeNaming) Namespace(req *restful.Request) (namespace string, err error) {
+	if n.scope.Name() == meta.RESTScopeNameRoot {
```
should this block only be done in the `if !ok || len(values) == 0 {` case? Otherwise a request to `foo?namespace=bar` is treated as namespace none

lavalamp replied:
Does this break events, which can either have a namespace or not depending on what they are for?

smarterclayton replied:
Good question, will dig in more.

----- Original Message -----
> > @@ -689,6 +709,9 @@ var _ ScopeNamer = legacyScopeNaming{}
> >  
> >  // Namespace returns the namespace from the query or the default.
> >  func (n legacyScopeNaming) Namespace(req *restful.Request) (namespace
> >  string, err error) {
> > +	if n.scope.Name() == meta.RESTScopeNameRoot {
> 
> Does this break events, which can either have a namespace or not depending on
> what they are for?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7969/files#r29967847

smarterclayton replied:
Events are namespaced scoped, so all three combinations are valid (empty and default to "default", set, and empty and no default applied).

The larger issue which I won't fix here is that we have a few other resources that are namespaced AND root scoped - events, and in openshift the subject and resource access reviews (which you can ask "can I do this" which involves actions in and out of namespaces).  For right now, they are listed as namespaced actions and api.NamespaceAll is a valid option for them.  We probably need to more formally define resources that support NamespaceAll, since there are other actions beyond "get collection" you might want to expose on a namespace all resource (post on the root to create a root event).

liggitt commented on pkg/kubectl/resource/visitor.go:
```diff
@@ -441,6 +441,9 @@ func FilterNamespace(info *Info) error {
 // set. If info.Object is set, it will be mutated as well.
 func SetNamespace(namespace string) VisitorFunc {
 	return func(info *Info) error {
+		if !info.Namespaced() {
```
should this err if `!info.Namespaced() && len(namespace) > 0`?

smarterclayton replied:
No, namespace is always provided into the builder, but this is specifically "set namespace if it can be set", not "hey, I expect set namespace to work on every object I pass it".

----- Original Message -----
> > @@ -441,6 +441,9 @@ func FilterNamespace(info *Info) error {
> >  // set. If info.Object is set, it will be mutated as well.
> >  func SetNamespace(namespace string) VisitorFunc {
> >  	return func(info *Info) error {
> > +		if !info.Namespaced() {
> 
> should this err if `!info.Namespaced() && len(namespace) > 0`?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7969/files#r29959580

derekwaynecarr commented:
@liggitt got here first, but I agree with his comments and otherwise LGTM

bgrant0607 commented:
cc @lavalamp

lavalamp commented on pkg/master/master.go:
```diff
@@ -671,7 +671,7 @@ func (m *Master) api_v1beta2() *apiserver.APIGroupVersion {
 	storage := make(map[string]rest.Storage)
 	for k, v := range m.storage {
-		if k == "minions" {
+		if k == "minions" || k == "minions/status" {
```
Is there a general problem to be solved with strings.Contains(k, "/")?

smarterclayton replied:
It would probably take a regex - I don't want minions or any of their sub resources.

----- Original Message -----
> > @@ -671,7 +671,7 @@ func (m *Master) api_v1beta2()
> > *apiserver.APIGroupVersion {
> >  func (m *Master) api_v1beta3() *apiserver.APIGroupVersion {
> >  	storage := make(map[string]rest.Storage)
> >  	for k, v := range m.storage {
> > -		if k == "minions" {
> > +		if k == "minions" || k == "minions/status" {
> 
> Is there a general problem to be solved with strings.Contains(k, "/")?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7969/files#r29967768

lavalamp replied:
ah-- gotcha. nvm then.

smarterclayton committed:
Allow v1beta3 to POST events to all namespaces

A namespaced resource that supports ALL may allow creation
on the root (all namespaces) collection, thus adding POST
here.

We need to better formalize the definition of calls on namespaced
resources at the root scope, so Storage objects that do not support
that call pattern can do so at definition time and reject those
calls.

smarterclayton commented:
Updated with comments, and added one more which correctly allows `POST /api/v1beta3/events` to be called (so that an event can be posted without a namespace).
